### ISSUE TITLE
Remove reports and recipe journals, replace with ILogger logging

### ISSUE THREAD

**Issue by DanielStolt** [OrchardCMS/Orchard#5491]
As per discussions with Sebastien yesterday, hereby propose to kill:

- The "reports" concept - it's not used for anything except logging SQL statements from migrations and logging one line from recipe execution. Also, it is hard-coded to the file system and does not work for web farms.
- Recipe journals - they provide no useful information at all, because they log at the step level.

These concept should be removed and replaced with good detailed `ILogger` logging from within all recipe-related components. This will make it a lot easier to troubleshoot recipe execution. Default Log4net configuration should be updated to log everything within the `Orchard.Recipes` namespace with level `INFO` to a separate logfile named `orchard-recipes-yyyy.mm.dd.log`.

@gcsuk has already done some great work to provide more detailed recipe logging, but based on recipe journals (which unfortunately does not work 100% because recipe journals are written in batches per step and only flushed out to disk if the whole step succeeds). As part of this, we would take the logging code he has added and just "rebase" it on ILogger instead.

Work to be performed in branch `feature/recipelogging`.

In the future, instead of the reports concept, we may elect to resurrect some form of "admin-consumable log accessible through the admin UI" but this would then be based on the `ILogger` abstraction.

**Comment by MpDzik** [comment]
Good idea, +1

**Comment by CSurieux** [comment]
The report concept was interesting but its content should reside in DB.
It contains a potential of user 'friendliness' where logs are technical.
Let's imagine that a customer buy something and you have a large/long process until the fulfillment, report could be interesting.

**Comment by Piedone** [comment]
Disagree on removing reporting and agree with @CSurieux with having it in the DB.

It is a useful feature that is underused in Orchard, but that under-usage doesn't lie in the nature of the feature itself (see: http://english.orchardproject.hu/blog/features-of-orchard-you-ve-never-used-reports). Specifically it's useful because it's "admin-consumable" and it's not a log: it's intended for users, not for developers. Although similar from a helicopter view, it's not a log and shouldn't be treated as a (technical) log.

**Comment by DanielStolt** [comment]
As I've understood it (read it somewhere) moving it to the database is tricky because the transaction might need to be rolled back if the request fails, and then anything logged will be rolled back also. True or false?

**Comment by sfmskywalker** [comment]
True, the DB does not seem to be the right place to store log entries because of that reason.

**Comment by CSurieux** [comment]
Maybe not writing the DB inside the running sole HTTP/NH transaction. This is a reccuring requirement.

**Comment by sfmskywalker** [comment]
@Piedone perhaps "Reports" isn't the right term for such a feature then. Reports sounds more like something that displays data in an easy to read manner. Also I'd argue that the way it was currently used is very user friendly in terms of content. Perhaps it makes more sense for Reports to display log entries of a specific log level or within a specific category. If not, what would be good examples of use cases where you want to create messages for just the user?

**Comment by Piedone** [comment]
@DanielStolt I'd vote false :-). I've written an implementation of IReportsPersister that stores reports in the DB as part of the Lombiq Hosting Suite: it uses a separate work context and is flushed even if there are exceptions.

@sfmskywalker probably "reports" is misleading, I agree, so we could rename it to something else. Do I really need to list examples where you want to notify admins in a persisted way about something happening in the system? :-) E-mails successfully sent out, indexing all content items done, invoices generated for recent purchases (Media Theme deployment succeeded/failed on DotNest)...

**Comment by sfmskywalker** [comment]
@Piedone thank you for those examples. :-)  Just making sure I understand what you're saying. Wouldn't you say that is still some kind of log? If so, couldn't we leverage the same logging infrastructure but specify a different category or channel? Then the "Reports" screen would simply pull from the correct source.

Regarding your IReportsPersister implementation, that would work of course if you never rollback the transaction. But the nice thing about ILogger and the Log4Net implementation is that you can use various appenders, including one, but not limited to, appenders that log to a database.

**Comment by DanielStolt** [comment]
@Piedone agree with your examples, and agree that this functionality **would** be nice to have **if** it were actually used. :) What @sfmskywalker is saying is what @sebastienros suggested also - to use ILogger for this, use the infrastructure we have in place, all the logging frameworks and abstractions, separate this kind of "user oriented" logging by a different category, channel it to a different target, and add functionality for surfacing that data in the admin UI.

**Comment by Piedone** [comment]
@DanielStolt I'm actually using it :-).

The fundamental issue I see here is that although we have everything in place for *writing* logs there is nothing to *read* them: and this is OK now as logs are generally not consumed by the same app that writes them (e.g. you use log parsers to handle log files, or log entries are sent to an entirely different, external system). This is not something simple to solve: if you read in log files it will cause a lot of scalability issues (big log files or in multi-node setups the need to replicate or share them), if you use an external system (like Azure Application Insights) you'd need to write a consumer for it (if it has an API).

So although in principle I wouldn't oppose storing a user-facing notification stream and technical logs in the same kind of "database" (although it doesn't seem "right" and I can see issues where people able to access technical logs will now gain access to private information intended for admins) I fail to see the benefit of building a new system for the same purpose.

**Comment by DanielStolt** [comment]
@Piedone OK so we have an infrastructural component, you are using it, but Orchard itself does not - then one might reasonably argue that perhaps you should simply move it into your hosting suite? :)

**Comment by Piedone** [comment]
Yes, I'm guilty of using one of Orchard's features :-). But why would we remove it exactly? Does it need maintenance? Does it negatively affects something? Do we know whether people are using it or not?

**Comment by sfmskywalker** [comment]
I don't know if it's used by anyone, but I've never seen anyone talk about it on the forums, except for [Bertrand in 2011](http://orchard.codeplex.com/workitem/17242) ;-)

**Comment by DanielStolt** [comment]
@Piedone I was only (half) joking with that comment. Jokes aside, I do believe that we should not have infrastructural services in the code base which are not actually used in the code base. In general, I'm sure you agree that *all* code carries with it an implicit cost burden? Maintenance, performance, learning curve, complexity, etc etc. You can always argue for any individual piece of code that its cost is small, there is no magic line between acceptable and unacceptable cost, which is why all we can do is apply a general principle to remove unused code and continuously keep the code base as lean and clean as possible.

**Comment by Piedone** [comment]
Maybe there are no issues because it's already perfect? :-D And yes theoretically there is an implicit cost with having any code in the code base though probably it's not Reports that makes learning Orchard hard or Orchard a big effort to maintain (not being touched in years: https://github.com/OrchardCMS/Orchard/tree/dev/src/Orchard.Web/Core/Reports).

Look guys we've probably debated too much on this matter and it was my fault to side track this issue on the question whether to keep reports or not. I illustrated why I think they're useful, even if arguably nothing in Orchard itself uses them in a suitable way. I also illustrated what the issues are with using the logging infrastructure, so IMO it really comes down to keep it or remove it, not to change it.

But since it's a Core module I wouldn't take removing it lightly (or generally removing any feature without serious consideration: Orchard is a platform that is used in ways you'd never imagine): I'd suggest a vote of the Steering Committee if you bring it up, and if we vote to remove it then we can make it deprecated as with other modules and then finally remove it in a later version.

**Comment by sfmskywalker** [comment]
I think that's a good idea, let's put it to a vote.

**Comment by dcinzona** [comment]
I was looking through the reports module a while back (over a year ago) because I wanted to use it to do some logging that could be digested via the Admin UI.  It seems useful, and simpler to manage than creating an audit trail or custom logging implementation.  Our scenario right now would be reports for updates to the datawarehouse, etc.  I don't think those need to be stored as audit logs or as error logs because, the thought is, they are neither.  The implementation of reports is also pretty simple, from what I'm reading right now

That being said, the only thing I see potentially being an issue is with the create function.  Should it create or return the id of the report if the report already exists?  

Right now, if I have it create a report in my code, and I have multiple instances running, isn't there a potential for duplicates?

```
        public int CreateReport(string title, string activityName) {
            lock ( _synLock ) {
                LoadReports();
                _isDirty = true;
                var reportId = _reports.Count == 0 ? 1 : _reports.Max(r => r.ReportId) + 1;
                var report = new Report {ActivityName = activityName, ReportId = reportId, Title = title, Utc = DateTime.UtcNow};
                _reports.Add(report);
                return reportId;
            }
        }
```
Instead of having a blanket "create", perhaps do an internal CreateOrGet function and have create call that.  That way no one needs to update their code to support CreateGet.

```
        public int CreateReport(string title, string activityName) {
            return CreateOrGetReport(title, activityName); 
        }

        public int CreateOrGetReport(string title, string activityName) {
            lock ( _synLock ) {
                LoadReports();
                var report = _reports.Where(x => String.Equals(x.Title, title) && String.Equals(x.ActivityName, activityName)).FirstOrDefault();
                if (report != null)
                    return report.ReportId;
               _isDirty = true;
                var reportId = _reports.Count == 0 ? 1 : _reports.Max(r => r.ReportId) + 1;
                report = new Report { ActivityName = activityName, ReportId = reportId, Title = title, Utc = DateTime.UtcNow };
                _reports.Add(report);
                return reportId;
            }
        }
```

Or maybe I'm missing something and there's no issue with the possibility of creating multiple reports with the same name/activity because it's handled somewhere else.

**Comment by dcinzona** [comment]
Ok, so no need to implement the code above... figured out the whole thing when I added it to my code.  
Other than that, seems like reports could benefit from trimming and search...like the much more robust audit module.  If only audit log records were as easy to implement as reports... literally, like 2 lines of code (generate, and log).

As a general rule, I think trimming should be supported by any method that automatically and repeatedly writes to a database.

**Comment by DanielStolt** [comment]
After today's triage meeting, I think we can all agree on the following compromise:

1. Remove the half-measure logging attempts during recipe execution (reports and recipe journals) and replace with detail robust ILogger logging (sent to a dedicated logfile as per above).
2. Keep the `Reports` feature but deprecate it (both because it is not used anywhere in Orchard AND because it has some serious drawbacks such as lack of farm support). This allows Zombiq and others to keep using it if they want.
3. Build something better for admin-viewable "logs" or "reports", something that supports structured entries and properly supports farm scenarios.

1 and 2 will be included in my branch. 3 to be done at any time thereafter.

**Comment by sfmskywalker** [comment]
:+1:

**Comment by DanielStolt** [comment]
Have not deprecated `Reports` for now, since that has yet to be voted on, but I have implemented everything else. Merging this into **dev** now.

**Comment by DanielStolt** [comment]
Fixed in bfe2c296ec454c1c677813de03ffae5e789af7ad.

**Comment by Piedone** [comment]
Steering Committee members please take a look at this thread, we'll vote on deprecating Reports on Tuesday. @bleroy @pszmyd @sebastienros @sfmskywalker
